clusterloader2/pkg/measurement/common/slos/api_responsiveness_prometheus.go

```diff
@@ -0,0 +1,284 @@
+	latencySamples, err := gatherSamples(c, timeBoundedLatencyQuery, measurementEnd)
+	if err != nil {
+		return nil, err
+	}
+	timeBoundedCountQuery := fmt.Sprintf(countQuery, measurementutil.ToPrometheusTime(measurementDuration))
+	countSamples, err := gatherSamples(c, timeBoundedCountQuery, measurementEnd)
+	if err != nil {
+		return nil, err
+	}
+	return a.convertToApiCalls(latencySamples, countSamples)
+}
+
+func (a *apiResponsivenessMeasurementPrometheus) convertToApiCalls(latencySamples, countSamples []*model.Sample) ([]apiCall, error) {
+	apiCalls := make(map[string]*apiCall)
+	ignoredResources := sets.NewString("events")
+	// TODO(krzysied): figure out why we're getting non-capitalized proxy and fix this.
+	ignoredVerbs := sets.NewString("WATCH", "WATCHLIST", "PROXY", "proxy", "CONNECT")
+
+	for _, sample := range latencySamples {
+		resource := string(sample.Metric["resource"])
+		subresource := string(sample.Metric["subresource"])
+		verb := string(sample.Metric["verb"])
+		scope := string(sample.Metric["scope"])
+		quantile, err := strconv.ParseFloat(string(sample.Metric["quantile"]), 64)
+		if err != nil {
+			return nil, err
+		}
+		if ignoredResources.Has(resource) || ignoredVerbs.Has(verb) {
+			continue
+		}
+
+		latency := time.Duration(float64(sample.Value) * float64(time.Second))
+		addLatency(apiCalls, resource, subresource, verb, scope, quantile, latency)
+	}
+
+	for _, sample := range countSamples {
+		resource := string(sample.Metric["resource"])
+		subresource := string(sample.Metric["subresource"])
+		verb := string(sample.Metric["verb"])
+		scope := string(sample.Metric["scope"])
+		if ignoredResources.Has(resource) || ignoredVerbs.Has(verb) {
+			continue
+		}
+
+		count := int(math.Round(float64(sample.Value)))
+		addCount(apiCalls, resource, subresource, verb, scope, count)
+	}
+
+	var result []apiCall
+	for _, call := range apiCalls {
+		result = append(result, *call)
+	}
+	return result, nil
+}
+
+func getApiCall(apiCalls map[string]*apiCall, resource, subresource, verb, scope string) *apiCall {
+	key := getMetricKey(resource, subresource, verb, scope)
+	call, exists := apiCalls[key]
+	if !exists {
+		call = &apiCall{
+			Resource:    resource,
+			Subresource: subresource,
+			Verb:        verb,
+			Scope:       scope,
+		}
+		apiCalls[key] = call
+	}
+	return call
+}
+
+func addLatency(apiCalls map[string]*apiCall, resource, subresource, verb, scope string, quantile float64, latency time.Duration) {
+	call := getApiCall(apiCalls, resource, subresource, verb, scope)
+	call.Latency.SetQuantile(quantile, latency)
+}
+
+func addCount(apiCalls map[string]*apiCall, resource, subresource, verb, scope string, count int) {
+	if count == 0 {
+		return
+	}
+	call := getApiCall(apiCalls, resource, subresource, verb, scope)
+	call.Count = count
+}
+
+func gatherSamples(c clientset.Interface, query string, queryTime time.Time) ([]*model.Sample, error) {
```

Comment:
This is basically executing prometheus query and parsing it.
I'm confident there will be many measurements that will like to do something exactly like this.

Could you move it to some util package and rename the method and make it public?

Reply:
Can we do this in separate PR? Current PR is already huge :/
I will create an issue regarding this change.

Reply:
Sure, thanks!

Reply:
Issue created https://github.com/kubernetes/perf-tests/issues/470